### PR TITLE
Convert Publisher and Subscription to use C++ Classes

### PR DESCRIPTION
**Pull request by ahcorde**
Part of #665.

This converts the Publisher and Subscription functions to a C++ classes. 

I have some issue with the QoSEvent destructor in some of the tests:

```bash
Thread 29 "python3" received signal SIGSEGV, Segmentation fault.
[Cambiando a Thread 0x7fffa3fff700 (LWP 338069)]
0x00000000006152cf in _is_legal_capsule (invalid_capsule=0x6d7160 "PyCapsule_GetPointer called with invalid PyCapsule object", capsule=0x3) at ../Objects/capsule.c:84
84	../Objects/capsule.c: No existe el archivo o el directorio.
(gdb) bt
#0  0x00000000006152cf in _is_legal_capsule (invalid_capsule=0x6d7160 "PyCapsule_GetPointer called with invalid PyCapsule object", capsule=0x3) at ../Objects/capsule.c:84
#1  PyCapsule_GetPointer (o=<unknown at remote 0x3>, name=0x7ffff5985881 "rcl_node_t") at ../Objects/capsule.c:84
#2  0x00007ffff6c182fb in rclpy_handle_get_pointer_from_capsule () from /home/ahcorde/ros2_galactic/install/lib/librclpy_common.so
#3  0x00007ffff5932510 in rclpy::Handle::rcl_ptr(char const*) () from /home/ahcorde/ros2_galactic/build/rclpy/test_rclpy/_rclpy_pybind11.cpython-38-x86_64-linux-gnu.so
#4  0x00007ffff5918bc3 in rcl_node_t* rclpy::Handle::cast_or_warn<rcl_node_t*>(char const*) () from /home/ahcorde/ros2_galactic/build/rclpy/test_rclpy/_rclpy_pybind11.cpython-38-x86_64-linux-gnu.so
#5  0x00007ffff593f2c4 in rclpy::Publisher::Publisher(pybind11::capsule, pybind11::object, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, pybind11::object)::{lambda(rcl_publisher_t*)#1}::operator()(rcl_publisher_t*) const [clone .lto_priv.0] () from /home/ahcorde/ros2_galactic/build/rclpy/test_rclpy/_rclpy_pybind11.cpython-38-x86_64-linux-gnu.so
#6  0x00007ffff594567e in std::_Sp_counted_deleter<rcl_publisher_t*, rclpy::Publisher::Publisher(pybind11::capsule, pybind11::object, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, pybind11::object)::{lambda(rcl_publisher_t*)#1}, std::allocator<void>, (__gnu_cxx::_Lock_policy)2>::_M_dispose() [clone .lto_priv.0] ()
   from /home/ahcorde/ros2_galactic/build/rclpy/test_rclpy/_rclpy_pybind11.cpython-38-x86_64-linux-gnu.so
#7  0x00007ffff58f464e in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release() () from /home/ahcorde/ros2_galactic/build/rclpy/test_rclpy/_rclpy_pybind11.cpython-38-x86_64-linux-gnu.so
#8  0x00007ffff58edfbb in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count() () from /home/ahcorde/ros2_galactic/build/rclpy/test_rclpy/_rclpy_pybind11.cpython-38-x86_64-linux-gnu.so
#9  0x00007ffff58ec990 in std::__shared_ptr<rcl_publisher_t, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr() ()
   from /home/ahcorde/ros2_galactic/build/rclpy/test_rclpy/_rclpy_pybind11.cpython-38-x86_64-linux-gnu.so
#10 0x00007ffff58ec9d8 in std::shared_ptr<rcl_publisher_t>::~shared_ptr() [clone .lto_priv.0] () from /home/ahcorde/ros2_galactic/build/rclpy/test_rclpy/_rclpy_pybind11.cpython-38-x86_64-linux-gnu.so
#11 0x00007ffff588c75c in rclpy::QoSEvent::~QoSEvent() [clone .lto_priv.0] () from /home/ahcorde/ros2_galactic/build/rclpy/test_rclpy/_rclpy_pybind11.cpython-38-x86_64-linux-gnu.so
#12 0x00007ffff5953d36 in rclpy::QoSEvent::~QoSEvent() () from /home/ahcorde/ros2_galactic/build/rclpy/test_rclpy/_rclpy_pybind11.cpython-38-x86_64-linux-gnu.so
#13 0x00007ffff5953d76 in std::default_delete<rclpy::QoSEvent>::operator()(rclpy::QoSEvent*) const () from /home/ahcorde/ros2_galactic/build/rclpy/test_rclpy/_rclpy_pybind11.cpython-38-x86_64-linux-gnu.so
#14 0x00007ffff588bc24 in std::unique_ptr<rclpy::QoSEvent, std::default_delete<rclpy::QoSEvent> >::~unique_ptr() [clone .lto_priv.0] ()
   from /home/ahcorde/ros2_galactic/build/rclpy/test_rclpy/_rclpy_pybind11.cpython-38-x86_64-linux-gnu.so
#15 0x00007ffff594c117 in pybind11::class_<rclpy::QoSEvent, rclpy::Destroyable>::dealloc(pybind11::detail::value_and_holder&) ()
   from /home/ahcorde/ros2_galactic/build/rclpy/test_rclpy/_rclpy_pybind11.cpython-38-x86_64-linux-gnu.so
#16 0x00007ffff589ac2f in pybind11::detail::clear_instance(_object*) () from /home/ahcorde/ros2_galactic/build/rclpy/test_rclpy/_rclpy_pybind11.cpython-38-x86_64-linux-gnu.so
#17 0x00007ffff589ad30 in pybind11_object_dealloc () from /home/ahcorde/ros2_galactic/build/rclpy/test_rclpy/_rclpy_pybind11.cpython-38-x86_64-linux-gnu.so
```
Do you mind to have a look @hidmic or @sloretz ?

Signed-off-by: ahcorde <ahcorde@gmail.com>

### CODE REVIEW

File: rclpy/src/rclpy/publisher.cpp [hunk 1]
```diff
@@ -75,60 +48,62 @@ publisher_create(
...
-  if (!pub) {
-    throw std::bad_alloc();
-  }
-  pub->publisher = rcl_get_zero_initialized_publisher();
-  pub->node = node;
+  rcl_publisher_ = std::shared_ptr<rcl_publisher_t>(
+    new rcl_publisher_t,
+    [this](rcl_publisher_t * publisher)
+    {
+      if (!node_handle_) {
```
Comment (hidmic) [inline rclpy/src/rclpy/publisher.cpp:55]: @ahcorde how can this ever happen? IIRC the whole purpose of `Handle` is to ensure ordered destruction.

File: rclpy/src/rclpy/publisher.hpp [hunk 2]
```diff
@@ -17,87 +17,111 @@
...
-void
-publisher_publish_raw(py::capsule pypublisher, std::string msg);
+class Publisher : public Destroyable
+{
+public:
+  /// Create a publisher
+  /**
+   * This function will create a publisher and attach it to the provided topic name
+   * This publisher will use the typesupport defined in the message module
+   * provided as pymsg_type to send messages over the wire.
```
Comment (hidmic) [inline rclpy/src/rclpy/publisher.hpp:40]: @ahcorde nit: I'd rather not talk about "wire". We're careful elsewhere not to imply anything about the transport.  
```suggestion
   * provided as pymsg_type to send messages.
```

File: rclpy/src/rclpy/publisher.hpp [hunk 3]
```diff
@@ -17,87 +17,111 @@
...
+  /// Get rcl_client_t pointer
+  rcl_publisher_t *
+  rcl_ptr() const
+  {
+    return rcl_publisher_.get();
+  }
+
+  /// Get rcl_client_t pointer
+  std::shared_ptr<rcl_publisher_t>
+  get_shared_ptr()
```
Comment (hidmic) [inline rclpy/src/rclpy/publisher.hpp:110]: @ahcorde nit: 
```suggestion
  get_rcl_shared_ptr()
```

File: rclpy/src/rclpy/subscription.cpp [hunk 4]
```diff
@@ -81,19 +54,35 @@ subscription_create(
...
-  if (!sub) {
-    throw std::bad_alloc();
-  }
-  sub->subscription = rcl_get_zero_initialized_subscription();
-  sub->node = node;
+  rcl_subscription_ = std::shared_ptr<rcl_subscription_t>(
+    new rcl_subscription_t,
+    [this](rcl_subscription_t * subscription)
+    {
+      if (!node_handle_) {
```
Comment (hidmic) [inline rclpy/src/rclpy/subscription.cpp:61]: @ahcorde same question about destruction order guarantees.

File: rclpy/test/test_destruction.py [hunk 5]
```diff
@@ -111,11 +111,7 @@ def test_destroy_subscription_asap():
             with sub.handle:
                 pass
 
-            with sub.handle:
-                node.destroy_subscription(sub)
-                # handle valid because it's still being used
-                with sub.handle:
-                    pass
+            node.destroy_subscription(sub)
```
Comment (hidmic) [inline rclpy/test/test_destruction.py:114]: @ahcorde hmm, this test seems valid to me, no? Same elsewhere.

**Review by hidmic — COMMENTED** [review]
> Thread 29 "python3" received signal SIGSEGV, Segmentation fault.

If I'm not mistaken, the issue is that `QoSEvent` no longer holds a `Handle` to an `rcl_subscription_t`, but the `rcl_subscription_t` itself. Destruction order w.r.t. your node handle is no longer guaranteed.

I believe the plan was to move away from `Handle` instances and simply use `std::shared_ptr` instances. Is that right @sloretz? If so, `QoSEvent` should keep an `std::shared_ptr` to its parent, which in turn keeps an `std::shared_ptr` to the node.

**Comment by sloretz** [comment]
> If I'm not mistaken, the issue is that QoSEvent no longer holds a Handle to an rcl_subscription_t, but the rcl_subscription_t itself. Destruction order w.r.t. your node handle is no longer guaranteed.
> 
> I believe the plan was to move away from Handle instances and simply use std::shared_ptr instances. Is that right @sloretz? If so, QoSEvent should keep an std::shared_ptr to its parent, which in turn keeps an std::shared_ptr to the node.

Ohhh shoot. I was thinking the `QoSEvent` just needed to keep an `std::shared_ptr` to the parent rcl type to keep that alive, but I forgot that doesn't keep it's grandparent, the `Node`, alive.

You're right that we'll need a shared pointer to all the ancestors. I'm not sure what that looks like to implement, but a shared pointer to the parent wrapper sounds reasonable.

File: rclpy/test/test_destruction.py (hunk 5 above)
Comment (ahcorde) [reply rclpy/test/test_destruction.py:114]: Refer to this comment: https://github.com/ros2/rclpy/pull/747#discussion_r606395943

File: rclpy/src/rclpy/publisher.cpp (hunk 1 above)
Comment (ahcorde) [reply rclpy/src/rclpy/publisher.cpp:55]: maybe there is some wrong with the destructor somewhere else, but If I remove this lines then the test_qos_event will fail

Comment (hidmic) [reply rclpy/src/rclpy/publisher.cpp:55]: Indeed. Default destructors for `Publisher` and `Subscription` are dropping `node_handle_` too early (and thus leaking!). Inverting member fields order or a custom destructor should do.

File: rclpy/src/rclpy/qos_event.hpp [hunk 6]
```diff
@@ -41,10 +43,10 @@ class QoSEvent : public Destroyable
    * Raises MemoryError if the event can't be allocated
    * Raises RCLError if event initialization failed in rcl
    *
-   * \param[in] pysubscription Capsule containing the subscription
+   * \param[in] subscriber Subscriber containing the subscription
```
Comment (hidmic) [inline rclpy/src/rclpy/qos_event.hpp:46]: @ahcorde consider: 
```suggestion
   * \param[in] subscription Subscription wrapping the underlying ``rcl_subscription_t`` object.
```

File: rclpy/src/rclpy/qos_event.hpp [hunk 7]
```diff
@@ -55,10 +57,10 @@ class QoSEvent : public Destroyable
    * Raises MemoryError if the event can't be allocated
    * Raises RCLError if event initialization failed in rcl
    *
-   * \param[in] pypublisher Capsule containing the publisher
+   * \param[in] publisher Publisher class
```
Comment (hidmic) [inline rclpy/src/rclpy/qos_event.hpp:60]: @ahcorde consider: 
```suggestion
   * \param[in] publisher Publisher wrapping the underlying ``rcl_publisher_t`` object.
```

File: rclpy/src/rclpy/publisher.cpp [hunk 8]
```diff
@@ -75,60 +48,52 @@ publisher_create(
...
-  if (!pub) {
-    throw py::error_already_set();
-  }
+  auto node = node_handle_->cast<rcl_node_t *>("rcl_node_t");
 
-  const char * node_logger_name = rcl_node_get_logger_name(pub->node);
+  const char * node_logger_name = rcl_node_get_logger_name(node);
   if (!node_logger_name) {
     throw RCLError("Node logger name not set");
   }
```
Comment (hidmic) [inline rclpy/src/rclpy/publisher.cpp:99]: @ahcorde hmm why does `Publisher::get_logger_name()` throw on error, whereas `Subscription::get_logger_name()` returns `None`? I'd rather be consistent.

File: rclpy/src/rclpy/subscription.hpp [hunk 9]
```diff
@@ -17,74 +17,98 @@
...
+   * This subscription will use the typesupport defined in the message module
+   * provided as pymsg_type to send messages.
+   *
+   * Raises RCLError if the subscription could not be created
+   *
+   * \param[in] pynode Capsule pointing to the node to add the subscriber to
+   * \param[in] pymsg_type Message module associated with the subscriber
+   * \param[in] topic The topic name
+   * \param[in] pyqos_profile rmw_qos_profile_t object for this subscription
+   * \return capsule containing the rclpy_subscription_t
```
Comment (hidmic) [inline rclpy/src/rclpy/subscription.hpp:48]: @ahcorde 
```suggestion

```
?

File: rclpy/src/rclpy/subscription.hpp [hunk 10]
```diff
@@ -17,74 +17,98 @@
...
- *   Metadata is a plain dictionary.
- */
-py::object
-subscription_take_message(py::capsule pysubscription, py::object pymsg_type, bool raw);
+  /// Get rcl_client_t pointer
+  std::shared_ptr<rcl_subscription_t>
+  get_rcl_shared_ptr()
+  {
+    return rcl_subscription_;
+  }
```
Comment (hidmic) [inline rclpy/src/rclpy/subscription.hpp:100]: @ahcorde unused?

File: rclpy/src/rclpy/utils.cpp [hunk 11]
```diff
@@ -104,20 +104,10 @@ get_rmw_implementation_identifier()
 }
 
 void
-assert_liveliness(py::capsule pyentity)
+assert_liveliness(rclpy::Publisher publisher)
```
Comment (hidmic) [inline rclpy/src/rclpy/utils.cpp:107]: @ahcorde better not to pass by value:
```suggestion
assert_liveliness(rclpy::Publisher * publisher)
```

**Comment by ahcorde** [comment]
Building up-to `rclpy` and testing `rclpy`

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14286/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9075/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11970/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14388/)

**Comment by ahcorde** [comment]
Thank for retargeting the PR, merging!

**Pull request merged**